packages/anchor-listener/src/listener.ts

```diff
@@ -0,0 +1,152 @@
+import type { SupportedNetwork } from '@ceramicnetwork/anchor-utils'
+import type { Provider } from '@ethersproject/providers'
+import {
+  type Observable,
+  type OperatorFunction,
+  type RetryConfig,
+  filter,
+  fromEvent,
+  map,
+  pipe,
+  scan,
+} from 'rxjs'
+
+import { type BlockProofs, mapLoadBlocksProofs } from './loader.js'
+
+export type BlockEvent = BlockProofs & {
+  reorganized: false
+}
+
+export type ReorganizedBlockEvent = BlockProofs & {
+  reorganized: true
+  expectedParentHash: string
+}
+
+export type BlockProofsListenerEvent = BlockEvent | ReorganizedBlockEvent
+
+export type ListenerParams = {
+  /* ethers.js Provider */
+  provider: Provider
+  /* supported anchor network chain ID */
+  chainId: SupportedNetwork
+  /* number of confirmations to wait for to load blocks */
+  confirmations: number
+  /* optional parent hash expected when loading the first block */
+  expectedParentHash?: string
+  /* optional Rx retry config */
+  retryConfig?: RetryConfig
+}
+
+type OrderingState = { continuous: Array<number>; latestContinuous?: number; maxFuture?: number }
+
+/**
+ * Create an Observable of arrays of continuous block numbers, ensuring ordering
+ *
+ * @param provider ethers.js provider
+ * @param confirmations optional number of confirmations, defaults to 0
+ * @returns Observable<Array<number>>
+ */
+export function createContinuousBlocksListener(
+  provider: Provider,
+  confirmations = 0
+): Observable<Array<number>> {
+  return fromEvent(provider, 'block').pipe(
+    // Map to latest block number minus wanted confirmations
+    map((blockNumber) => (blockNumber as number) - confirmations),
+    // Order continuous block numbers, keeping track of highest block received
+    scan<number, OrderingState, OrderingState>(
+      (state, blockNumber) => {
+        if (state.latestContinuous != null) {
+          if (blockNumber <= state.latestContinuous) {
+            // Already processed block number
+            return { ...state, continuous: [] }
+          }
+          if (blockNumber > state.latestContinuous + 1) {
+            // Future block number, keep track
+            return {
```

Comment:
FWIW, I think this is what I was trying to get at with my comment here: https://github.com/ceramicnetwork/js-ceramic/pull/2599#discussion_r1052660029

I feel like it would be perfectly safe to start processing blocks as soon as we hear about them, no need to wait.  But happy to defer to Steph and Paul to figure out if it makes sense to wait to make reorg handling easier.

Reply:
Yes I agree that doesn't really make sense anymore, I originally thought we wouldn't provide the blocks and anchors when reorganizations are detected so I wanted to avoid overfetching, but we need to provide them anyways so we can simplify the process, I'll update the logic.